src/ucp/tag/rndv.c

```diff
@@ -29,6 +29,88 @@ static int ucp_tag_rndv_is_get_put_op_possible(ucp_ep_h ep, ucp_lane_index_t lan
     }
 }
 
+static void ucp_tag_rndv_resolve_lanes(ucp_request_t *req)
+{
+    ucp_ep_h ep       = req->send.ep;
+    uint64_t lane_map = -1;
+    ucp_rndv_get_lane_info_t *info;
+    ucp_lane_index_t lane;
+    ucp_rsc_index_t i;
+
+    UCS_STATIC_ASSERT(UCP_MAX_LANES < (sizeof(lane_map) * 8));
+
+    ucs_assert(!ucp_ep_is_stub(ep));
+    ucs_assert_always(req->send.rndv_get.num_lanes <= UCP_MAX_RNDV_LANES);
+    ucs_assert(req->send.rndv_get.rkey != NULL);
+
+    /* already resolved? ok, just exit */
+    if (req->send.rndv_get.rkey->resolved) {
+        return;
+    }
+
+    /* lookup lanes for known memory domains */
+    for (i = 0; i < req->send.rndv_get.num_lanes; /* no increment here */) {
+        info = ucp_tag_rndv_get_lane_info(req, i);
+        ucs_assert(info->lane == UCP_NULL_LANE);
+        ucs_assert(info->md_index != UCP_NULL_RESOURCE);
+        for (lane = 0; lane < ucp_ep_num_lanes(ep); lane++) {
+            if ((lane_map & UCS_BIT(lane)) &&
+                (info->md_index == ucp_ep_md_index(ep, lane))) {
+                info->lane = lane;
+                lane_map &= ~UCS_BIT(lane); /* do not allow repeat lanes */
+                break;
+            }
+        }
+
+        if (info->lane != UCP_NULL_LANE) {
+            i++;
+        } else {
+            /* could not find pair lane - release rkey */
+            if (ucp_tag_rndv_is_rkey_valid(req, i)) {
+                uct_rkey_release(ucp_tag_rndv_rkey_bundle(req, i));
+            }
+            /* shift rest of rkeys to current location */
+            if (i + 1 < req->send.rndv_get.num_lanes) {
+                memmove(info, ucp_tag_rndv_get_lane_info(req, i + 1),
+                        (req->send.rndv_get.num_lanes - i - 1) * sizeof(ucp_rndv_get_lane_info_t));
+            }
+            req->send.rndv_get.num_lanes--;
+        }
+    }
+
+    req->send.rndv_get.rkey->resolved = 1;
+}
+
+static int ucp_tag_rndv_is_get_op_possible(ucp_request_t *req)
+{
+    ucp_ep_h ep = req->send.ep;
+    uint64_t md_flags;
+    ucp_lane_index_t i;
+    ucp_lane_index_t lane;
+
+    ucs_assert(!ucp_ep_is_stub(ep));
+
+    if (req->send.rndv_get.rkey == NULL) {
+        return 0;
+    }
+
+    ucp_tag_rndv_resolve_lanes(req);
+
+    if (req->send.rndv_get.num_lanes == 0) {
+        return 0;
+    }
+
+    for (i = 0; i < req->send.rndv_get.num_lanes; i++) {
+        lane = ucp_tag_rndv_get_lane_info(req, i)->lane;
+        md_flags = ucp_ep_md_attr(ep, lane)->cap.flags;
+        if ((md_flags & UCT_MD_FLAG_NEED_RKEY) && !ucp_tag_rndv_is_rkey_valid(req, i)) {
```

Comment:
you already checked that in `ucp_tag_rndv_resolve_lanes`. Seems no need on this check

Reply:
resolve_lanes just looks for corresponding lane, it doesn't check for flags & rkey valid value. in case if something wrong (for example tag offloading failed to provide rkey in future) - rndv will safely fallback to AM mode

Reply:
can't we do that in `ucp_tag_rndv_resolve_lanes`? Otherwise why would we need to add lanes which can't be used for rndv?

Reply:
ok, fixed